### DISCUSSION TITLE
Subscribing with a bit of history

### DISCUSSION THREAD

**Discussion by nielsbasjes** [graphql-java/graphql-java#3404] [Q&A]
I'm working on a homeautomation hobby project. I have events from a physical device (some kind of sensor; Solar converter, heatpump, electricity meter, ...) and I want to create an app (Flutter) that shows me a `realtime` indicator of the current value and a graph with the values for the last few hours (which is also updated on each newly arriving event).

Because the physical device itself speaks modbus I want to create a GraphQL api over that to make it more practically usable.

For the graph in my app I need to have the latest value available and a bit of history (say few hours worth) to ensure the graph is filled when the user opens the app.

I now have 
- a separate subscription (i.e. only the latest events as they arrive)
- a separate query for the history.

I'm sure I can get this working; yet I had this idea on how it may be easier to use:

Q: Is it possible to have `a new subscriber` receive a limited set of `old events` before they receive the current events in the stream?

I want to have the subscriber optionally indicate how much history they want (either number of events or earliest timestamp field in the event record) and indicate the granularity of the events (like not all but only 1 per minute or an average per minute).

So only this specific new subscriber will get this history and only what this subscriber indicated they want.

Is this possible to combine this into the outgoing stream for this single specific subscriber?
Or is a separate subscription and a separate query the only way to do this?

**Comment by bbakerman** [comment]
> Is it possible to have a new subscriber receive a limited set of old events before they receive the current events in the stream?

This is less a question on the graphql subscription and more a question for the backing implementation of events.

For example does the "store" that has the events have this capability?  If you are using Reactor say for the JVM api to get events are you using the methods that allow access to past events?

**Reply by nielsbasjes** [reply]
The key thing is my current limited understanding of the available backing implementations and what the real capabilities are. I'm really new in this area.

In the end I want to persist this data to a backend (i.e. some kind of database) to have as much history as I think is useful.

I have now created my first very basic subscription implementation (using reactor.core.publisher.Flux) that just provides the latest value and does not yet store it anywhere  :

https://github.com/nielsbasjes/dsmr-tools/tree/main/dsmr-graphql/src/main/java/nl/basjes/dsmr/graphql

This reads data from a physical device (in this case an electricity meter) and publishes that to all who want it.

Right now my experimental code stores a bit of history in memory (only accessible via a query) and in the Flux (accessible via a subscription)

What I'm looking for is a basic example implementation (any backing implementation) that allows a new subscriber access the available history as a starting point.
